File: Data Structures/Binary Search Tree/BST Validation/Validation.cpp
#include <iostream>
#include <queue>
#include <climits>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    Node(int value) : data(value), left(nullptr), right(nullptr) {}
};

Node* buildTree() {
    int value;
    cout << "Enter the root node value: ";
    cin >> value;
    Node* root = new Node(value);

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        Node* current = q.front();
        q.pop();

        cout << "Enter left child of " << current->data << " (-1 for no child): ";
        cin >> value;
        if (value != -1) {
            current->left = new Node(value);
            q.push(current->left);
        }

        cout << "Enter right child of " << current->data << " (-1 for no child): ";
        cin >> value;
        if (value != -1) {
            current->right = new Node(value);
            q.push(current->right);
        }
    }

    return root;
}

bool inorderTraversal(Node* root, int& prevValue) {
    if (root == nullptr) {
        return true;
    }

    if (!inorderTraversal(root->left, prevValue)) {
        return false;
    }

    if (root->data <= prevValue) {
        return false;
    }

    prevValue = root->data;

    return inorderTraversal(root->right, prevValue);
}

bool isBST(Node* root) {
    int prevValue = INT_MIN;
    return inorderTraversal(root, prevValue);
}

int main() {
    Node* root = buildTree();

    if (isBST(root)) {
        cout << "The tree is a valid Binary Search Tree (BST)." << endl;
    } else {
        cout << "The tree is NOT a valid Binary Search Tree (BST)." << endl;
    }

    return 0;
}

File: Data Structures/Binary Search Tree/BST Validation/Validation.py
class Node:
    def __init__(self, value):
        self.data = value
        self.left = None
        self.right = None

def buildTree():
    value = int(input("Enter the root node value: "))
    root = Node(value)

    queue = [root]
    while queue:
        current = queue.pop(0)

        value = int(input(f"Enter left child of {current.data} (-1 for no child): "))
        if value != -1:
            current.left = Node(value)
            queue.append(current.left)

        value = int(input(f"Enter right child of {current.data} (-1 for no child): "))
        if value != -1:
            current.right = Node(value)
            queue.append(current.right)

    return root

def inorderTraversal(root, prevValue):
    if root is None:
        return True

    if not inorderTraversal(root.left, prevValue):
        return False

    if root.data <= prevValue[0]:
        return False

    prevValue[0] = root.data

    return inorderTraversal(root.right, prevValue)

def isBST(root):
    prevValue = [float('-inf')]
    return inorderTraversal(root, prevValue)

if __name__ == "__main__":
    root = buildTree()

    if isBST(root):
        print("The tree is a valid Binary Search Tree (BST).")
    else:
        print("The tree is NOT a valid Binary Search Tree (BST).")
